sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java

```diff
@@ -0,0 +1,272 @@
+import org.apache.calcite.jdbc.JavaTypeFactoryImpl;
+import org.apache.calcite.prepare.CalciteCatalogReader;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataTypeSystem;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.schema.ScalarFunction;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.schema.impl.ScalarFunctionImpl;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperatorTable;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.util.ChainedSqlOperatorTable;
+import org.apache.calcite.tools.FrameworkConfig;
+import org.apache.calcite.tools.Frameworks;
+import org.apache.calcite.tools.Planner;
+
+import com.datatorrent.api.DAG;
+
+/**
+ * SQL Execution Environment for SQL integration API of Apex.
+ * This exposes calcite functionality in simple way. Most of the APIs are with builder pattern which makes it
+ * easier to construct a DAG using this object.
+ *
+ * Eg.
+ * <p>
+ * SQLExecEnvironment.getEnvironment(dag)
+ *                   .registerTable("TABLE1", < Object of type {@link Endpoint} >)
+ *                   .registerTable("TABLE2", < Object of type {@link Endpoint} >)
+ *                   .executeSQL("INSERT INTO TABLE2 SELECT STREAM * FROM TABLE1);
+ * </p>
+ *
+ * Above code will evaluate SQL statement and convert the resultant Relational Algebra to a sub-DAG.
+ */
+public class SQLExecEnvironment
+{
+  private static boolean driverLoaded = false;
+
+  /**
+   * Calcite driver needs to be loaded statically.
+   * When the application requiring calcite integration is launched from apex cli, calcite driver is not loaded as
+   * static block of calcite Driver class has not for called.
+   * This is the entry class for calcite SQL DAG. Hence this static block will always gets called before anything else.
+   */
+  static {
+    loadCalciteDriver();
+  }
+
+  private final JavaTypeFactoryImpl typeFactory;
+  private SchemaPlus schema = Frameworks.createRootSchema(true);
+  private DAG dag;
+
+  /**
+   * Construct SQL Execution Environment which works on given DAG objec.
+   *
+   * @param dag LogicalPlan object that will be populated with sub-DAG for SQL operations.
+   */
+  private SQLExecEnvironment(DAG dag)
+  {
+    this.dag = dag;
+    this.typeFactory = new JavaTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+  }
+
+  /**
+   * Given SQLExec{@link SQLExecEnvironment} object for given {@link DAG}.
+   *
+   * @param dag Object of type {@link DAG}
+   * @return Returns {@link SQLExecEnvironment} object
+   */
+  public static SQLExecEnvironment getEnvironment(DAG dag)
+  {
+    return new SQLExecEnvironment(dag);
+  }
+
+  private static void loadCalciteDriver()
+  {
+    if (!driverLoaded) {
+      try {
+        Class.forName("org.apache.calcite.jdbc.Driver");
```

Comment:
Are you sure that classes are loaded on demand?

Try taking out that block and see what happens. You might be surprised.

I see you get the connection using DriverManager.getConnection. That is the standard way to connect to JDBC, and should be safe. See http://docs.oracle.com/javase/6/docs/api/java/util/ServiceLoader.html, which is used by JDBC.